### PR TITLE
PERF: to_numeric for numeric dtypes

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Skip `object` conversion if input is numeric already.

```
-  146.41ms    26.45μs      0.00  miscellaneous.to_numeric.time_from_float
```

### CODE REVIEW

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jreback** [comment]
```
In [1]: pd.date_range('20130101',periods=3,tz='US/Eastern')
Out[1]: DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00', '2013-01-03 00:00:00-05:00'], dtype='datetime64[ns, US/Eastern]', freq='D')

In [2]: pd.to_numeric(pd.date_range('20130101',periods=3,tz='US/Eastern'))
TypeError: 'Index' does not have the buffer interface

In [3]: pd.to_numeric(pd.date_range('20130101',periods=3,tz='US/Eastern').to_series())
Out[3]: 
2013-01-01 00:00:00-05:00    1357016400000000000
2013-01-02 00:00:00-05:00    1357102800000000000
2013-01-03 00:00:00-05:00    1357189200000000000
Freq: D, dtype: int64
```

**Comment by jreback** [comment]
@sinhrks this looked fine. can you just rebase  (I think this was sent when codecov was misbehaving).

**Comment by jreback** [comment]
we can push this as well, lmk.

**Comment by sinhrks** [comment]
@jreback Current master has following issues. I've updated the PR to included the fixes.

#### 1. ``to_numeric(Index)`` returns ``np.ndarray``, not ``Index``

To be compat with ``to_datetime``, it must be ``Index``.

```
pd.to_numeric(pd.Index([1, 2, 3]))
# array([1, 2, 3])
```

#### 2. datetime-likes are not actually supported

Based on your example, it should be ``asi8`` repr.

```
pd.to_numeric(pd.date_range('2011-01-01', freq='M', periods=3))
# TypeError: 'Index' does not have the buffer interface

pd.to_numeric(pd.Series(pd.date_range('2011-01-01', freq='M', periods=3)))
# TypeError: Invalid object type
```

File: pandas/tools/tests/test_util.py [hunk 1]
```diff
@@ -145,6 +188,83 @@ def test_type_check(self):
...
+        res = pd.to_numeric(idx.values)
+        tm.assert_numpy_array_equal(res, idx.asi8)
+
+    def test_period(self):
+        idx = pd.period_range('2011-01', periods=3, freq='M', name='xxx')
+        res = pd.to_numeric(idx)
+        tm.assert_index_equal(res, pd.Index(idx.asi8, name='xxx'))
+
+        # ToDo: enable when we can support native PeriodDtype
+        # res = pd.to_numeric(pd.Series(idx, name='xxx'))
```
Comment (sinhrks) [inline pandas/tools/tests/test_util.py]: Avoided to use `is_period_array` ATM. we can have faster impl when period dtype is added.

File: pandas/tools/util.py [hunk 2]
```diff
@@ -78,29 +78,44 @@ def to_numeric(arg, errors='raise'):
...
+        values = arg.asi8
+        if values is None:
+            values = arg.values
     elif isinstance(arg, (list, tuple)):
-        arg = np.array(arg, dtype='O')
+        values = np.array(arg, dtype='O')
     elif getattr(arg, 'ndim', 1) > 1:
         raise TypeError('arg must be a list, tuple, 1-d array, or Series')
+    else:
+        values = arg
```
Comment (jreback) [inline pandas/tools/util.py]: I think this will fail with a scalar (though that might be another issue).

Comment (sinhrks) [inline pandas/tools/util.py]: Currently it fails. Will include the fix.

```
pd.to_numeric(1)
# ValueError: Buffer has wrong number of dimensions (expected 1, got 0)
```

Comment (jreback) [inline pandas/tools/util.py]: gr8!. I *think* there might be an issue about this somewhere.

Comment (jreback) [inline pandas/tools/util.py]: https://github.com/pydata/pandas/issues/12136#issuecomment-200349724

not a separate issue.

**Comment by sinhrks** [comment]
Added scalar impl and tests (``test_scalar``). Now green except for ``codecov/changes``.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**